libraries/AP_Mount/AP_Mount_Topotek.cpp

```diff
@@ -0,0 +1,995 @@
+        break;
+    case '2':
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "%s started tracking", send_message_prefix);
+        break;
+    }
+}
+
+// gimbal distance information analysis
+void AP_Mount_Topotek::gimbal_dist_info_analyse(void)
+{
+    if ('E' == _msg_buff[10] && 'R'== _msg_buff[11] && 'R'==_msg_buff[12]) {
+        _measure_dist_m = -1.0f;
+        return;
+    }
+    long lrf = (char_to_number(_msg_buff[10])*100000 +
+                char_to_number(_msg_buff[11])*10000 +
+                char_to_number(_msg_buff[12])*1000 +
+                char_to_number(_msg_buff[13])*100 +
+                char_to_number(_msg_buff[14])*10 +
+                char_to_number(_msg_buff[16]));
+
+    _measure_dist_m = lrf/10 + lrf%10 * 0.1;
+}
+
+// gimbal basic information analysis
+void AP_Mount_Topotek::gimbal_basic_info_analyse(void)
+{
+    uint8_t major_ver = 0;
+    uint8_t minor_ver = 0;
+    uint8_t patch_ver = 0;
+
+    // analyze gimbal model and firmware version
+    for (uint8_t count = 0; count < char_to_number(_msg_buff[5]); count++) {
+        if (count == 0) {
+            major_ver = char_to_number(_msg_buff[10 + count]);
+        } else if (count == 1) {
+            minor_ver = char_to_number(_msg_buff[10 + count]);
+        } else if (count == 2) {
+            patch_ver = char_to_number(_msg_buff[10 + count]);
+        } else {
+            strncpy((char*)_model_name, (const char*)(_msg_buff+13), 10);
+            break;
+        }
+    }
+    _firmware_ver = (patch_ver << 16) | (minor_ver << 8) | (major_ver);
+
+    if (strlen((const char*)_model_name) == 0) {
+        strcpy((char*)_model_name, "Unknown");
+    }
+
+    // display gimbal model and firmware version to user
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Mount: Topotek %s v%u.%u.%u",
+        _model_name,
+        major_ver,
+        minor_ver,
+        patch_ver);
+
+    _got_gimbal_basic_info = true;
+}
+
+// add check
+void AP_Mount_Topotek::add_check(int8_t *cmd, uint8_t len)
+{
+    int8_t crc = 0;
+    crc = calculate_crc(cmd, len);
+    cmd[len] = hex2char((crc >> 4) & 0x0f);
+    cmd[len+1] = hex2char(crc & 0x0f);
+}
+
+// calculate checksum
+int8_t AP_Mount_Topotek::calculate_crc(int8_t *cmd, uint8_t len)
+{
+    uint8_t crc = 0;
+    for (uint16_t i = 0; i<len; i++) {
+        crc += cmd[i];
+    }
+    return(crc);
+}
+
+// hexadecimal to character conversion
+int8_t AP_Mount_Topotek::hex2char(int8_t data)
+{
+    if ((9 >= data)) {
+        return (data + '0');
+    } else {
+        return (data - 10 + 'A');
+    }
+}
+
+// character to number conversion
+uint8_t AP_Mount_Topotek::char_to_number(int8_t data)
+{
+    if (('0' <= data && '9' >= data)) {
+        return (data - '0');
+    }
+    else if (('a' <= data) && ('z' >= data)) {
+        return (data - 'a' + 10);
+    }
+    else if (('A' <= data) && ('Z' >= data)) {
+        return (data - 'A' + 10);
+    }
+    return 0;
+}
+
```

Comment:
These methods are all available elsewhere in ArduPilot

Reply:
Thanks @peterbarker, I've replaced char-to-number with AP_Common's char-to-hex.  I could not find an equivalent to hex2char though.  I initially thought that AP_Common's hex_to_uint8() was the same but it's actually for a completely different purpose as far as I can tell.